### DISCUSSION TITLE
how can I generate the user-defined MAC address from hash of IKE identity?

### DISCUSSION THREAD

**Discussion by f1-outsourcing** [strongswan/strongswan#2165] [Q&A]
How should I generate these mac addresses so I can configure them in the dhcp server? Maybe good to put as foot note on the manual[1]
https://docs.strongswan.org/docs/5.9/plugins/dhcp.html

**Comment by tobiasbrunner** [comment]
Not easy to do (it's a 32-bit SipHash-2-4 of the identity with a [static key](https://github.com/strongswan/strongswan/blob/6f8275ababa99238887071d175ca76cd3801fa35/src/libstrongswan/utils/chunk.c#L1029) plus the prefix `0x7aa7`). Might be easier to let clients connect one time and then add mappings. Otherwise, map the identity directly, it's sent as client-identifier DHCP option (see [this integration test](https://strongswan.org/testing/testresults/ikev2/dhcp-static-client-id/)).

**Reply by f1-outsourcing** [reply]
I must be blind or so, but I only see this option `identity_lease = yes` on (see [this integration test](https://strongswan.org/testing/testresults/ikev2/dhcp-static-client-id/)).

**Reply by tobiasbrunner** [reply]
Check the DHCP config on venus in [console.log](https://strongswan.org/testing/testresults/ikev2/dhcp-static-client-id/console.log).

**Reply by f1-outsourcing** [reply]
I have even logging on the dhcpd, but I don't see anything logged except the standard. 'dhcp-client-identifier' does not seem to be send.

```
log(info, concat(">dhcp-client-identifier = ", option dhcp-client-identifier));

dhcpd[4091921]: >dhcp-client-identifier = 45704477-eb23-4dd9-9294-451d4e806113eth1
dhcpd[4091921]: >dhcp-client-identifier = 45704477-eb23-4d
```
(I know this is not correct (binary), but still I get something logged)

**Reply by tobiasbrunner** [reply]
A request from strongSwan is received, though? And did you enable `identity_lease`?

**Reply by f1-outsourcing** [reply]
Yes I get even an ip address from a range that I should not be getting. Maybe this[1] would be nice to add as feature request. Being able to set the vendor-class-identifier so I can redirect everything from the vpn server to a specific pool. 

identity_lease is on because I am logging a totally different mac address starting with a 7

[1]
https://github.com/strongswan/strongswan/discussions/1994#discussioncomment-8774364

**Reply by tobiasbrunner** [reply]
> Yes I get even an ip address from a range that I should not be getting.

If the request doesn't contain a client identifier, I'd assume you configured `identity_lease` incorrectly (the only other reason would be that the identity is too long, but there should be enough room for more than 200 characters). Also note that if the identities are FQDNs, the plugin adds a host name option (12) containing the identity (it does so even if `identity_lease` is disabled).

> Maybe this[1] would be nice to add as feature request. Being able to set the vendor-class-identifier so I can redirect everything from the vpn server to a specific pool.

Maybe, you're free to file a feature request (but note that it probably won't have that much priority at the moment). If you're using ISC's dhcpd, you should also be able to put the strongSwan requests into a separate class based on the MAC address with something like `match if substring (hardware,1,2) = 7a:a7;`.

**Comment by f1-outsourcing** [comment]
> A request from strongSwan is received, though? And did you enable `identity_lease`?

You are right commenting this identity_lease  out, does not change the behaviour. So I guess this file is not being loaded.
```
charon {
        plugins {
                dhcp {
                        #identity_lease = yes
                }

                kernel-netlink {
                        mtu = 1400
                        #mss = 1481
                }
        }
}
```

**Reply by tobiasbrunner** [reply]
In which file did you configure this? And on client or server?

**Reply by f1-outsourcing** [reply]
on server /etc/swanctl/conf.d/charon-custom.conf

**Reply by f1-outsourcing** [reply]
oops putting it here, gives me the dhcp identification. Sorry about this. 
/etc/strongswan.d/charon-custom.conf

**Comment by f1-outsourcing** [comment]
I think I have found a bug. I have two connection profiles 
ikev2 rsa gives me this client-identifier 

`>dhcp-client-identifier = 0"1 0#036#006#003U#004#003#023#027aaaa.bbbb.cccc..com `

and ikev2 eap-tls gives me a correct client-identifier

`>dhcp-client-identifier = aaaa.bbbb.cccc.com`

**Reply by f1-outsourcing** [reply]
linux client is ok, only have this with windows client

**Reply by tobiasbrunner** [reply]
It's not a bug. The identity passed to the DHCP server is the identity sent by the client as-is. In case of Windows it apparently sends the full subject DN of the certificate (that's the binary ASN.1 encoding), so that's what you receive at the DHCP server.

**Reply by f1-outsourcing** [reply]
Oh interesting, I thought strongswan was adding this.

**Reply by tobiasbrunner** [reply]
A distinguished name always starts with `0x30` (ASN.1 SEQUENCE tag), which is why you see that string starting with `0` (ASCII encoding of `0x30`). So it might be possible to pre-process the client identity in DHCPD. For instance, match on that first byte and then use `substring()` to strip the prefix from such identities (and leave the others that don't start with `0x30` as they are). Depending on the identities you use, the prefix might always be the same, or at least its length doesn't change. You seem to use `CN=...` as subject DNs and the hex encoding of the example you posted is `0x30223120301e06035504031317`, followed by the actual value (the last byte is the length of that value, here `0x17` = 23 - that's 27 in octal encoding, by the way, which is what you see in the ASCII output as `#027` in the log). So that's basically an offset of 14 characters that you could pass to `substring()` to get at the actual value.
